Tbyfield/dc 2025 04 01 preview

tbyfield opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

tbyfield committed:
Copy files from stable/2025-02-01

Copied the files in a separate commit.
This allows reviewers to easily diff subsequent changes against the previous spec.

tbyfield committed:
Update version to preview/2025-04-01-preview

Updated the API version from stable/2025-02-01 to preview/2025-04-01-preview.

tbyfield committed:
Added tag for 2025-04-01-preview in readme file

tbyfield committed:
2025-04-01-preview changes

- re-add encryptionSet resource. This was previously in 2024-10-01-preview but had been removed for 2025-02-01 (stable) as it wasn't ready
- added image definition resource under devcenter catalog scope. These are read-only resources (same as the existing one under project catalog route) and these resources are sync from files in a connected git repo (which the catalog is associated with)
- Added additional properties to Project to allow customers to enable/disable several new features
* UserCustomizations - configures when the users can use user level customizations when creating a dev box
* DevBoxAutoDelete - configures auto-deletion after set time and grace period
* Azure AI services integration which is create managed resources for AI development
* Serverless GPU which will create managed resources to allow developers to open a remote terminal with GPU access
* Workspace Storage which is will create a storage associated with a Dev Box and Serverless GPU
* Active hours configuration
* Ability to enable/disable Dev Box tunnel access for a Pool

tbyfield commented:
DevCenter 2025-04-01-preview changes

- re-add encryptionSet resource. This was previously in 2024-10-01-preview but had been removed for 2025-02-01 (stable) as it wasn't ready
- added image definition resource under devcenter catalog scope. These are read-only resources (same as the existing one under project catalog route) and these resources are sync from files in a connected git repo (which the catalog is associated with)
- Added additional properties to Project to allow customers to enable/disable several new features 
* UserCustomizations - configures when the users can use user level customizations when creating a dev box
* DevBoxAutoDelete - configures auto-deletion after set time and grace period
* Azure AI services integration which is create managed resources for AI development
* Serverless GPU which will create managed resources to allow developers to open a remote terminal with GPU access
* Workspace Storage which is will create a storage associated with a Dev Box and Serverless GPU
* Active hours configuration 
* Ability to enable/disable Dev Box tunnel access for a Pool

razvanbadea-msft commented:
Fix required checks

razvanbadea-msft commented:
>       "post": {

shouldnt this be a get

---
Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:4742 in 7af38ed. [](commit_id = 7af38eda1ae1759b4bb371c25ddca090ee3d88a8, deletion_comment = False)

razvanbadea-msft commented:
>       "post": {

same here about get

---
Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:5013 in 7af38ed. [](commit_id = 7af38eda1ae1759b4bb371c25ddca090ee3d88a8, deletion_comment = False)

razvanbadea-msft commented:
>           "$ref": "../../../../../common-types/resource-management/v4/managedidentity.json#/definitions/ManagedServiceIdentity"

why are you not using v5

---
Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:5571 in 7af38ed. [](commit_id = 7af38eda1ae1759b4bb371c25ddca090ee3d88a8, deletion_comment = False)

razvanbadea-msft commented:
>         "keyEncryptionKeyUrl": {

you can also add url format to it

---
Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:5599 in 7af38ed. [](commit_id = 7af38eda1ae1759b4bb371c25ddca090ee3d88a8, deletion_comment = False)

razvanbadea-msft commented:
>           "type": "string",

you can add the arm-id format

---
Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:5930 in 7af38ed. [](commit_id = 7af38eda1ae1759b4bb371c25ddca090ee3d88a8, deletion_comment = False)

razvanbadea-msft commented:
>           "description": "The condition that needs to be met to run the task."

Could you expand on this description a bit? This will be used to generate comments & documentation, so it is used by customers to discover what this property mean.

---
Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:7740 in 7af38ed. [](commit_id = 7af38eda1ae1759b4bb371c25ddca090ee3d88a8, deletion_comment = False)

razvanbadea-msft commented:
Fix required checks

tbyfield committed:
PR comment updates

- Updated ManagedServiceIdentity common-type version from v4 to v5 for DevCenterEncryptionSet
- Added uri format to keyEncryptionUrl
- Added arm-id format to identityResourceId

tbyfield committed:
Add suppression

tbyfield commented:
> Fix required checks

The Avocado checks are failing due to this known issue: https://github.com/Azure/avocado/issues/150. 

We previously removed some routes prior to our Stable (2025-02-01) and have deprecated them so not bringing them back.

tbyfield committed:
Update description of condition to make it a little clearer.

tbyfield commented:
> > ```
> >   "post": {
> > ```
> 
> shouldnt this be a get
> 
> Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:4742 in [7af38ed](https://github.com/Azure/azure-rest-api-specs/commit/7af38eda1ae1759b4bb371c25ddca090ee3d88a8). [](commit_id = [7af38ed](https://github.com/Azure/azure-rest-api-specs/commit/7af38eda1ae1759b4bb371c25ddca090ee3d88a8), deletion_comment = False)

It's intentionally a POST. If it were a GET, we'd have to model it as a resource (constrained by ARM's rules) which doesn't make much too much sense. We didn't want to put it on the resource itself for a few reason, but one of the main ones was it potentially contains customer content. This had been the advice we were given for other catalog resources - i.e. environmentDefinitions and tasks that follow this pattern. Hence we'd want to keep it consistent with those

tbyfield committed:
undo update to ManagedServiceIdentity to v5. This needs to be consistent across all references and could potentially be breaking to existing resources

tbyfield commented:
> > ```
> >       "$ref": "../../../../../common-types/resource-management/v4/managedidentity.json#/definitions/ManagedServiceIdentity"
> > ```
> 
> why are you not using v5
> 
> Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:5571 in [7af38ed](https://github.com/Azure/azure-rest-api-specs/commit/7af38eda1ae1759b4bb371c25ddca090ee3d88a8). [](commit_id = [7af38ed](https://github.com/Azure/azure-rest-api-specs/commit/7af38eda1ae1759b4bb371c25ddca090ee3d88a8), deletion_comment = False)

Tried updating just this one, but that broke SDK generation (since there are other references to v4 this resulted in duplicate of same type). We'd need to update all our existing references, and I'm not 100% certain that wouldn't cause any issues to existing customers. So, I would prefer to keep this consistent and re-look at updating all in a later version.

tbyfield committed:
Merge branch 'main' into tbyfield/dc_2025-04-01-preview

tbyfield commented:
/azp run

gary-x-li commented:
>         },

There are many properties defined as a "boolean enum" with two values Enabled and Disabled. It's no better than a boolean since it's hard to extend in the future; and customer may question why not just defined them as boolean types.

Consider making it easier to extend by naming the object something like "DevBoxDeleteSettings" and the property "deleteMode". Then you can include values such as "Auto", "Manual", and more in the future.

Please go through all similar objects/properties.

---
Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:6203 in 758588a. [](commit_id = 758588a7a0eacf8bbd7733add024c6b09e227e58, deletion_comment = False)

gary-x-li commented:
>         },

There are many properties defined as a "boolean enum" with two values Enabled and Disabled. It's no better than a boolean since it's hard to extend in the future; and customer may question why not just defined them as boolean types.

Consider making it easier to extend by naming the object something like "DevBoxDeleteSettings" and the property "deleteMode". Then you can include values such as "Auto", "Manual", and more in the future.

Please go through all similar objects/properties.

---
Refers to: specification/devcenter/resource-manager/Microsoft.DevCenter/preview/2025-04-01-preview/devcenter.json:6203 in 758588a. [](commit_id = 758588a7a0eacf8bbd7733add024c6b09e227e58, deletion_comment = False)

tbyfield committed:
Addressing PR comments. Changed enablement enums from Enabled/Disabled to more meaningful values

tbyfield commented:
@gary-x-li thanks for the feedback. Agree with you on the enums. I've updated most of them inline with your suggestion. I will note this has been the guidance from many ARM reviewers to not use booleans and hence this is a general pattern we've adopted. 

The one that didn't really make sense was the userCustomizationsEnableStatus - this one most resembles a boolean on/off. I'd prefer to keep as is and consistent with the rest of our APIs ... but as long as this would be a consistent guidance I would not object to making it a bool

nickdepinet reviewed (approved):
ship it
